### PR TITLE
Improvements for correct functioning of the Driver OCI8

### PR DESCRIPTION
**Pull request by desenvuno**
- Record count of a Select object;
- Problem solved in the definition of variable names in Oracle Select object when used subselects - limit and offset

Reference: https://github.com/zendframework/zf2/pull/7444

### CODE REVIEW

**Comment by mwillbanks** [comment]
@ralphschindler care to take a look?

File: src/Adapter/Driver/Oci8/Oci8.php [hunk 1]
```diff
@@ -34,23 +41,40 @@ class Oci8 implements DriverInterface, Profiler\ProfilerAwareInterface
...
+            foreach ($features as $name => $feature) {
+                $this->addFeature($name, $feature);
+            }
+        } elseif ($features instanceof AbstractFeature) {
+            $this->addFeature($features->getName(), $features);
+        } elseif ($features === self::FEATURES_DEFAULT) {
+            $this->setupDefaultFeatures();
+        }
     }
-
```
Comment (Maks3w) [inline src/Adapter/Driver/Oci8/Oci8.php]: Undo all these cosmetic changes

File: test/Adapter/Driver/Oci8/ResultTest [hunk 2]
```diff
@@ -0,0 +1,92 @@
+<?php
```
Comment (Maks3w) [inline test/Adapter/Driver/Oci8/ResultTest]: missing file extension

File: src/Adapter/Driver/Oci8/Feature/RowCounter.php [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+ * @license   http://framework.zend.com/license/new-bsd New BSD License
+ */
+
+namespace Zend\Db\Adapter\Driver\Oci8\Feature;
+
+use Zend\Db\Adapter\Driver\Feature\AbstractFeature;
+use Zend\Db\Adapter\Driver\Oci8\Statement;
+
+/**
+ * RowCounter
```
Comment (Maks3w) [inline src/Adapter/Driver/Oci8/Feature/RowCounter.php]: Put a better description

File: src/Adapter/Driver/Oci8/Feature/RowCounter.php [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    /**
+     * @return string
+     */
+    public function getName()
+    {
+        return 'RowCounter';
+    }
+
+    /**
+     * @param \Zend\Db\Adapter\Driver\Oci8\Statement $statement
```
Comment (Maks3w) [inline src/Adapter/Driver/Oci8/Feature/RowCounter.php]: Replace with the aliased type

File: src/Adapter/Driver/Oci8/Oci8.php [hunk 5]
```diff
@@ -12,9 +12,16 @@
...
 use Zend\Db\Adapter\Exception;
 use Zend\Db\Adapter\Profiler;
+use Zend\Db\Adapter\Driver\Feature\AbstractFeature;
+use Zend\Db\Adapter\Driver\Feature\DriverFeatureInterface;
 
-class Oci8 implements DriverInterface, Profiler\ProfilerAwareInterface
+class Oci8 implements DriverInterface, DriverFeatureInterface, Profiler\ProfilerAwareInterface
 {
+     /**
+     * @const
```
Comment (Maks3w) [inline src/Adapter/Driver/Oci8/Oci8.php]: This tag is not necesasary, remove

File: src/Adapter/Driver/Oci8/Oci8.php [hunk 6]
```diff
@@ -182,19 +235,22 @@ public function createStatement($sqlOrResource = null)
...
     }
-
     /**
      * @param  resource $resource
-     * @param  null     $isBuffered
+     * @param  null     $context
      * @return Result
      */
-    public function createResult($resource, $isBuffered = null)
+    public function createResult($resource, $context = null)
```
Comment (Maks3w) [inline src/Adapter/Driver/Oci8/Oci8.php]: I'm not sure if this is a BC break for child classes

File: src/Adapter/Driver/Oci8/Result.php [hunk 7]
```diff
@@ -6,73 +6,66 @@
...
 class Result implements Iterator, ResultInterface
 {
     /**
      * @var resource
      */
     protected $resource = null;
-
     /**
-     * @var bool
+     * @var null
```
Comment (Maks3w) [inline src/Adapter/Driver/Oci8/Result.php]: Is this the only possible type?

File: src/Adapter/Driver/Oci8/Result.php [hunk 8]
```diff
@@ -6,73 +6,66 @@
...
-
     /**
      * Initialize
      * @param resource $resource
+     * @param null|int $generatedValue
+     * @param null|int $rowCount
      * @return Result
      */
-    public function initialize($resource /*, $generatedValue, $isBuffered = null*/)
+    public function initialize($resource, $generatedValue = null, $rowCount = null)
```
Comment (Maks3w) [inline src/Adapter/Driver/Oci8/Result.php]: Same about BC break

File: src/Adapter/Platform/Oracle.php [hunk 9]
```diff
@@ -9,12 +9,22 @@
...
 
+use Zend\Db\Adapter\Exception;
+use Zend\Db\Adapter\Driver\DriverInterface;
+use Zend\Db\Adapter\Driver\Oci8;
+use Zend\Db\Adapter\Driver\Pdo;
+
 class Oracle extends AbstractPlatform
 {
+
+    /** @var resource|\PDO */
```
Comment (Maks3w) [inline src/Adapter/Platform/Oracle.php]: null too

Comment (Maks3w) [inline src/Adapter/Platform/Oracle.php]: resource is not a type allowed in constructor neither setDriver. Also types mismatch with the assigned in setDriver

File: src/Adapter/Platform/Oracle.php [hunk 10]
```diff
@@ -9,12 +9,22 @@
...
+    protected $resource = null;
+
     /**
      * @param array $options
+     * @param null|\Zend\Db\Adapter\Driver\Oci8\Oci8|\Zend\Db\Adapter\Driver\Pdo\Pdo $driver
```
Comment (Maks3w) [inline src/Adapter/Platform/Oracle.php]: Use aliased types

File: src/Adapter/Platform/Oracle.php [hunk 11]
```diff
@@ -22,6 +32,36 @@ public function __construct($options = [])
...
+     * @param \Zend\Db\Adapter\Driver\Pdo\Pdo||\PDO||\Zend\Db\Adapter\Driver\Oci8\Oci8||\Oci8 $driver
+     * @throws \Zend\Db\Adapter\Exception\InvalidArgumentException
+     * @return $this
+     */
+    public function setDriver($driver)
+    {
+        if ($driver instanceof Oci8\Oci8
+            || ($driver instanceof Pdo\Pdo && $driver->getDatabasePlatformName() == 'Oracle')
+            || ($driver instanceof Pdo\Pdo && $driver->getDatabasePlatformName() == 'Sqlite')
+            || ($driver instanceof \oci8)
```
Comment (Maks3w) [inline src/Adapter/Platform/Oracle.php]: import type

File: src/Adapter/Platform/Oracle.php [hunk 12]
```diff
@@ -22,6 +32,36 @@ public function __construct($options = [])
...
+            || ($driver instanceof \PDO && $driver->getAttribute(\PDO::ATTR_DRIVER_NAME) == 'oci')
```
Comment (Maks3w) [inline src/Adapter/Platform/Oracle.php]: import type

File: src/Adapter/Platform/Oracle.php [hunk 13]
```diff
@@ -22,6 +32,36 @@ public function __construct($options = [])
...
+        ) {
+            $this->resource = $driver;
+            return $this;
+        }
+        throw new Exception\InvalidArgumentException('$driver must be a Oci8 or Oracle PDO Zend\Db\Adapter\Driver, Oci8 instance or Oci PDO instance');
```
Comment (Maks3w) [inline src/Adapter/Platform/Oracle.php]: Try to keep under 120 characters line length

File: src/Adapter/Platform/Oracle.php [hunk 14]
```diff
@@ -49,9 +89,22 @@ public function quoteIdentifierChain($identifierChain)
...
     {
+        if ($this->resource instanceof DriverInterface) {
+            $this->resource = $this->resource->getConnection()->getResource();
+        }
+
+        if ($this->resource) {
+            if ($this->resource instanceof \PDO) {
+                return $this->resource->quote($value);
+            }
+            if (get_resource_type($this->resource) == 'oci8 connection' || get_resource_type($this->resource) == 'oci8 persistent connection') {
```
Comment (Maks3w) [inline src/Adapter/Platform/Oracle.php]: Try to keep under 120 characters line length

File: src/Adapter/Platform/Oracle.php [hunk 15]
```diff
@@ -49,9 +89,22 @@ public function quoteIdentifierChain($identifierChain)
...
+                return '\'' . addcslashes(str_replace('\'', '\'\'', $value), "\x00\n\r\"\x1a") . '\'';
```
Comment (Maks3w) [inline src/Adapter/Platform/Oracle.php]: Use double quotes instead scape single quotes

File: src/Adapter/Driver/Oci8/Oci8.php (hunk 6 above)
Comment (weierophinney) [inline src/Adapter/Driver/Oci8/Oci8.php]: @Maks3w It's not an issue; the value is passed to `Zend\Db\Adapter\Driver\Oci8\Result`, which currently ignores it. Additionally, the interface defining it, `Zend\Db\Adapter\Driver\DriverInterface`, only defines the first argument, `$resource`. As such, the argument is completely unused at this point.

File: src/Adapter/Driver/Oci8/Result.php (hunk 8 above)
Comment (weierophinney) [inline src/Adapter/Driver/Oci8/Result.php]: See above. The body of the method was ignoring it already, and the argument is not defined in the governing interface (which is why it's an optional argument in the concrete definition).

File: src/Adapter/Driver/Oci8/Oci8.php (hunk 1 above)
Comment (weierophinney) [inline src/Adapter/Driver/Oci8/Oci8.php]: I'm pretty sure these are a result of running `php-cs-fixer` without specifying the configuration file. Unfortunately, `php-cs-fixer`, when run a second time *with* the configuration file, will not fix them. We can fix these on merge.

**Comment by romulobusatto** [comment]
We were in doubt regarding the comments and the bug label. We need to do more?

**Comment by weierophinney** [comment]
@romulobusatto Note - all merged! :-)

**Comment by romulobusatto** [comment]
Ok, thanks! :D

**Comment by romulobusatto** [comment]
I found a problem on the initialize of result in Oci8. I need do a new pull request?

**Pull request closed without merging**